base/strings/util.jl

```diff
@@ -711,3 +711,12 @@ julia> ascii("abcdefgh")
 ```
 """
 ascii(x::AbstractString) = ascii(String(x))
+
+rest(s::Union{String,SubString{String}}, i=1) = SubString(s, i)
```

Comment:
It might be worth using this implementation for all AbstractStrings by default. Are there any string types where the iteration state is not an index? Making that assumption seems worth it to avoid copying.

Reply:
I feel a bit uncomfortable making this assumption, since right now, I think we always regard the specific implementation of `iterate` as an implementation detail, at least for any abstract types. At least right now, we don't document this as part of any public interface and if anyone implemented `iterate` differently, I would imagine it would be quite hard to catch this. That said, I really don't know a lot about which types of custom string types people actually want to use.

Reply:
From @keno on Slack:
> ChainStrings tend to have iteration state that is (link, idx) pairs
> but need to index using character offsets still since we make that assumption

So I think we should keep this fallback as-is. It won't always be the most efficient, but I think performance should still be reasonable for a fallback method.